source/nijigenerate/viewport/vertex/mesheditor/tools/edgecutter.d

```diff
@@ -0,0 +1,120 @@
+import nijigenerate.viewport.common.mesh;
+import nijigenerate.viewport.base;
+import i18n;
+import nijigenerate.core.input;
+import nijigenerate.core.actionstack;
+import nijilive;
+import nijilive.core.dbg;
+import nijilive.math;
+import bindbc.imgui;
+import nijigenerate.widgets;
+
+class EdgeCutterTool : NodeSelect {
+    vec2 dragOrigin;
+    vec2 dragEnd;
+
+    override
+    void setToolMode(VertexToolMode toolMode, IncMeshEditorOne impl) {
+        isDragging = false;
+    }
+
+    override bool onDragStart(vec2 mousePos, IncMeshEditorOne impl) {
+        dragOrigin = mousePos;
+        dragEnd = mousePos;
+        isDragging = true;
+        return true;
+    }
+
+    override bool onDragUpdate(vec2 mousePos, IncMeshEditorOne impl) {
+        if (!isDragging) return false;
+        dragEnd = mousePos;
+        return true;
+    }
+
+    override bool onDragEnd(vec2 mousePos, IncMeshEditorOne impl) {
+        if (!isDragging) return false;
+        isDragging = false;
+        dragEnd = mousePos;
+
+        // Disconnect edges that intersect with the cutter line using impl interface (no casting)
+        auto action = impl.newMeshDisconnectAction(impl.getTarget().name);
+
+        vec2 a = dragOrigin;
+        vec2 b = dragEnd;
+
+        impl.forEachEdge((MeshVertex* v, MeshVertex* conn) {
+            // Disconnect only when the user-drawn segment [a,b] intersects edge [v,conn]
+            if (areLineSegmentsIntersecting(a, b, v.position, conn.position)) {
+                action.disconnect(v, conn);
+            }
+        });
+
+        action.updateNewState();
+        incActionPush(action);
+
+        impl.refreshMesh();
+        return true;
+    }
+
+    override bool update(ImGuiIO* io, IncMeshEditorOne impl, int action, out bool changed) {
+        super.update(io, impl, action, changed);
+
+        incStatusTooltip(_("Cut edges intersecting the line"), _("Drag Left Mouse"));
+
+        // Start on mouse down (no drag threshold), end on release
+        if (igIsMouseClicked(ImGuiMouseButton.Left)) {
+            onDragStart(impl.mousePos, impl);
+        }
+
+        if (isDragging && igIsMouseDown(ImGuiMouseButton.Left)) {
+            onDragUpdate(impl.mousePos, impl);
+        }
+
+        if (isDragging && incInputIsMouseReleased(ImGuiMouseButton.Left)) {
+            onDragEnd(impl.mousePos, impl);
+        }
+
+        return true;
+    }
+
+    override void draw(Camera camera, IncMeshEditorOne impl) {
+        super.draw(camera, impl);
+        if (!isDragging) return;
+        vec3[] lines = [vec3(dragOrigin, 0), vec3(dragEnd, 0)];
+        inDbgSetBuffer(lines);
+        inDbgLineWidth(3);
+        inDbgDrawLines(vec4(0, 0, 0, 1), mat4.identity());
+        inDbgDrawLines(vec4(1, 0, 0, 1), mat4.identity());
+    }
+}
+
+class ToolInfoImpl(T: EdgeCutterTool) : ToolInfoBase!(T) {
+    override
+    bool viewportTools(bool deformOnly, VertexToolMode toolMode, IncMeshEditorOne[Node] editors) {
```

Comment:
Please add check that any of target is sub-class of `Drawable`
(See `bezierdeform.d` for similar check.)
Same feedback applies to `canUse`